common/browser_process.go

```diff
@@ -166,28 +178,43 @@ func execute(
 		}
 	}()
 
-	return cmd, stdout, done, nil
+	return command{cmd, ctx, stdout, stderr}, nil
 }
 
 // parseDevToolsURL grabs the websocket address from chrome's output and returns it.
-func parseDevToolsURL(ctx context.Context, rc io.Reader) (wsURL string, _ error) {
+func parseDevToolsURL(cmd command) (wsURL string, _ error) {
 	type result struct {
 		devToolsURL string
 		err         error
 	}
 	c := make(chan result, 1)
 	go func() {
-		const prefix = "DevTools listening on "
+		const urlPrefix = "DevTools listening on "
+		errRx := regexp.MustCompile(`^\[.*:ERROR:.*\] (?P<errStr>.*)$`)
+		errTmpl := []byte("$errStr")
+		scanner := bufio.NewScanner(cmd.stderr)
 
-		scanner := bufio.NewScanner(rc)
 		for scanner.Scan() {
-			if s := scanner.Text(); strings.HasPrefix(s, prefix) {
+			line := scanner.Text()
+			if strings.HasPrefix(line, urlPrefix) {
 				c <- result{
-					strings.TrimPrefix(strings.TrimSpace(s), prefix),
+					strings.TrimPrefix(strings.TrimSpace(line), urlPrefix),
 					nil,
 				}
 				return
 			}
+			errM := []byte{}
+			// Golang's regexp module doesn't support backreferences, so do this
+			// workaround. We need to use regex, unfortunately, to extract the
+			// error message properly. It would be awkward to do otherwise, and
+			// this isn't a hot path, so it shouldn't impact performance.
+			for _, submatches := range errRx.FindAllSubmatchIndex([]byte(line), -1) {
+				errM = errRx.Expand(errM, errTmpl, []byte(line), submatches)
+			}
+			if len(errM) > 0 {
+				c <- result{"", errors.New(string(errM))}
+				return
+			}
```

Comment:
Since the Chromium errors have [defined patterns](https://source.chromium.org/chromium/chromium/src/+/main:base/logging.h;drc=4be70d0a54a18a36221e3733fd1dea827fc5f7ad;l=395) and never use `:ERROR:` in the rest of the message, could we use `strings` to extract the error?

```suggestion
		const urlPrefix = "DevTools listening on "
		scanner := bufio.NewScanner(cmd.stderr)

		for scanner.Scan() {
			line := scanner.Text()
			if strings.HasPrefix(line, urlPrefix) {
				c <- result{
					strings.TrimPrefix(strings.TrimSpace(line), urlPrefix),
					nil,
				}
				return
			}
			if strings.Contains(line, ":ERROR:") {
				if i := strings.Index(line, "] "); i > 0 {
					c <- result{"", errors.New(line[i+2:])}
					return
				}
			}
		}
```

Reply:
We could, but I still find the regex safer for parsing this. If this was a performance concern, I would agree, and would probably done it that way to begin with, but I think it's more important that we capture the relevant information correctly, than be a few microseconds faster.

Or do you have another reason to change this?

Reply:
My reasons were (IMHO) string matching is simpler, easy to read, probably slightly faster, and has less code.

> We could, but I still find the regex safer for parsing this.

So I'm not sure if the regex is safer 😒 For example, let's say the line contains the following:

```
"launching browser: Invalid devtools server [1020/141559.181070:ERROR:headless_shell.cc(771)] port"
```

* The regex captures the error as follows: `"launching browser: port"`
* However, the string match above captures it as this: `"launching browser: Invalid devtools server [1020/141559.181070:ERROR:headless_shell.cc(771)] port"`.

Reply:
It's totally up to you 😄

Reply:
Resolved by 64e2810.